src/shims/tls.rs

```diff
@@ -105,12 +105,25 @@ impl<'tcx> TlsData<'tcx> {
         thread_id: ThreadId,
         new_data: Scalar<Provenance>,
         cx: &impl HasDataLayout,
+        static_roots: &mut Vec<AllocId>,
     ) -> InterpResult<'tcx> {
         match self.keys.get_mut(&key) {
             Some(TlsEntry { data, .. }) => {
                 if new_data.to_target_usize(cx)? != 0 {
                     trace!("TLS key {} for thread {:?} stored: {:?}", key, thread_id, new_data);
                     data.insert(thread_id, new_data);
+                    if thread_id.to_u32() == 0 {
+                        if let Some(alloc) = (new_data.to_pointer(cx).ok())
+                            .and_then(|ptr| ptr.provenance.and_then(|prov| prov.get_alloc_id()))
+                        {
+                            trace!(
+                                "TLS key {} for main thread stored as static root: {:?}",
+                                key,
+                                alloc,
+                            );
+                            static_roots.push(alloc);
```

Comment:
This is definitely not what we want -- if a value is set for some TLS variable and then changed again, only the final value should be allowed to leak.

So if you want to also tackle library TLS state (which will be needed to make the test pass on Windows), that must happen around here

https://github.com/rust-lang/miri/blob/fcabf757701254a4d25e7ea6dc91f445f0eed14c/src/shims/tls.rs#L274

The `on_stack_empty` function should receive a flag indicating whether it should in the end leak or free the remaining thread-local state. [Here](https://github.com/rust-lang/miri/blob/8bbc93ca31f29b10168919f241d9c9be6b69f77e/src/eval.rs#L219) we'd pass in "leak" but [here](https://github.com/rust-lang/miri/blob/8bbc93ca31f29b10168919f241d9c9be6b69f77e/src/concurrency/thread.rs#L825) it should be "free".

This is a stretch goal if you want to tackle the Windows issue in the same PR. It's also fine to leave this to a future PR.

Reply:
Removed the partial implementation to leave Windows for a future PR